src/exception.c

```diff
@@ -409,19 +420,34 @@ void exc_fiq(u64 *regs)
     }
 
     if (cpu_features->fast_ipi) {
-        reg = mrs(SYS_IMP_APL_UPMCR0);
-        if ((reg & UPMCR0_IMODE_MASK) == UPMCR0_IMODE_FIQ &&
-            (mrs(SYS_IMP_APL_UPMSR) & UPMSR_IACT)) {
-            printf("  UPMC IRQ, masking\n");
-            reg_clr(SYS_IMP_APL_UPMCR0, UPMCR0_IMODE_MASK);
-        }
-
         if (mrs(SYS_IMP_APL_IPI_SR_EL1) & IPI_SR_PENDING) {
             printf("  Fast IPI IRQ, clearing\n");
             msr(SYS_IMP_APL_IPI_SR_EL1, IPI_SR_PENDING);
         }
     }
 
+    if (cpu_features->uncore_version) {
+        uint64_t upmcr0_imode = 0;
+        switch (cpu_features->uncore_version) {
+            case UNCORE_V1:
+                upmcr0_imode = UPMCR0_IMODE_T8015;
+                break;
+            case UNCORE_V2:
+                upmcr0_imode = UPMCR0_IMODE_T8020;
+                break;
+            default:
+                panic("Unhandled UPMC IRQ\n"); // FIQ storm is coming
```

Comment:
uh, how? I don't see how the default case can be reached. `UNCORE_NONE` would bail out earlier in the outer if

Reply:
this should be unreachable as long as there's no _V3 and similar